File: Linked Lists/sumLists.py
class Node(dict):
    def __init__(self, data=float('inf'), next=None):
        self.data = data
        self.next = next

def sumLists(linkedListOne={}, linkedListTwo={}):
    '''
    Solution - Traverse lists in parallel suming the nodes

    Complexity Analysis
    O(n) time | O(n) space

    Adds the two numbers and returns the sum as a linked list

    dict: linkedListOne
    dict: linkedListTwo
    return: Linked list with sumup numbers    
    '''
    # Gracefully handle type and Falsy values
    if (not isinstance(linkedListOne, dict) or bool(linkedListOne) == False or not isinstance(linkedListTwo, dict) or bool(linkedListTwo) == False):
        print('Arguments should be a valid non-empty dictionary')
        return False

    sumLinkedList = Node(0)
    currSumLinkedList = sumLinkedList

    currNodeOne = linkedListOne
    currNodeTwo = linkedListTwo

    carryOver = 0

    while (currNodeOne != None or currNodeTwo != None or carryOver > 0):
        nodeSum = 0

        if (currNodeOne):
            nodeSum += currNodeOne['data']
            currNodeOne = currNodeOne['next']

        if (currNodeTwo):
            nodeSum += currNodeTwo['data']
            currNodeTwo = currNodeTwo['next']

        nodeSum += carryOver
        carryOver = nodeSum // 10

        currSumLinkedList['next'] = dict({ 'data': nodeSum % 10, 'next': None })
        currSumLinkedList = currSumLinkedList['next']

    return sumLinkedList['next']

# Test cases (black box - unit testing)
testCases = [
    # Normal
    # Data that is typical (expected) and should be accepted by the system.
    { 
        'assert': sumLists({ 'data': 5, 'next': { 'data': 2, 'next': { 'data': 4, 'next': None } } }, { 'data': 3, 'next': { 'data': 8, 'next': { 'data': 1, 'next': None } } }),
        'expected': { 'data': 8, 'next': { 'data': 0, 'next': { 'data': 6, 'next': None } } }
    },
    { 
        'assert': sumLists({ 'data': 8, 'next': { 'data': 5, 'next': { 'data': 7, 'next': None } } }, { 'data': 2, 'next': { 'data': 9, 'next': { 'data': 0, 'next': None } } }),
        'expected': { 'data': 0, 'next': { 'data': 5, 'next': { 'data': 8, 'next': None } } }
    },
    { 
        'assert': sumLists({ 'data': 9, 'next': { 'data': 8, 'next': { 'data': 7, 'next': None } } }, { 'data': 4, 'next': { 'data': 9, 'next': { 'data': 0, 'next': { 'data': 8, 'next': None } } } }),
        'expected': { 'data': 3, 'next': { 'data': 8, 'next': { 'data': 8, 'next': { 'data': 8, 'next': None } } } }
    },
    { 
        'assert': sumLists(
            { 'data': 5, 'next': { 'data': 1, 'next': { 'data': 8, 'next': { 'data': 9, 'next': { 'data': 8, 'next': None }} } } },
            { 'data': 4, 'next': { 'data': 9, 'next': { 'data': 0, 'next': { 'data': 8, 'next': None } } } }
        ),
        'expected': { 'data': 9, 'next': { 'data': 0, 'next': { 'data': 9, 'next': { 'data': 7, 'next': { 'data': 9, 'next': None } } } } }
    },
    # Boundary data (extreme data, edge case)
    # Data at the upper or lower limits of expectations that should be accepted by the system.
    { 
        'assert': sumLists({ 'data': 1, 'next': None }, { 'data': 1, 'next': None }),
        'expected': { 'data': 2, 'next': None }
    },

    # Abnormal data (erroneous data)
    # Data that falls outside of what is acceptable and should be rejected by the system.
    { 'assert': sumLists(), 'expected': False },
    { 'assert': sumLists(0), 'expected': False },
    { 'assert': sumLists(''), 'expected': False },
    { 'assert': sumLists([]), 'expected': False },
    { 'assert': sumLists(()), 'expected': False },
    { 'assert': sumLists({}), 'expected': False },
    { 'assert': sumLists(None), 'expected': False },
    { 'assert': sumLists(False), 'expected': False }
]

# Run tests
for (index, test) in enumerate(testCases):
    print(f'# Test {index + 1}')
    print(f'Actual: {test["assert"]}')
    print(f'Expected: {test["expected"]}')
    print('🤘 Test PASSED 🤘' if test["assert"] == test["expected"] else '👎 Test FAILED 👎', '\n')

File: Linked Lists/sumLists.js
import { factoryLinkedListNode as createNode } from '../../Data Structures/linkedList.js';

/**
 * Solution - Traverse lists in parallel suming the nodes
 *
 * Complexity Analysis
 * O(n) time | O(n) space
 *
 * Adds the two numbers and returns the sum as a linked list
 *
 * @param {object} linkedListOne
 * @param {object} linkedListTwo
 * @returns {object} Linked list with sumup numbers
 */
function sumLists(linkedListOne = {}, linkedListTwo = {}) {
    // Gracefully handle type and Falsy values
    if (typeof linkedListOne !== 'object' || Object.keys(linkedListOne || {}).length === 0 || typeof linkedListTwo !== 'object' || Object.keys(linkedListTwo || {}).length === 0) {
        console.error('Parameters must be a valid non-empty object');
        return false;
    }

    const sumLinkedList = createNode({ data: 0 });
    let currSumLinkedList = sumLinkedList;

    let currNodeOne = linkedListOne;
    let currNodeTwo = linkedListTwo;

    let carryOver = 0

    while (currNodeOne !== null || currNodeTwo !== null || carryOver > 0) {
        let sum = 0;

        if (currNodeOne) {
            sum += currNodeOne.data;
            currNodeOne = currNodeOne.next;
        }

        if (currNodeTwo) {
            sum += currNodeTwo.data;
            currNodeTwo = currNodeTwo.next;
        }

        sum += carryOver;
        carryOver = Math.floor(sum / 10);

        currSumLinkedList.next = createNode({ data: Number(sum % 10) });
        currSumLinkedList = currSumLinkedList.next;
    }

    return sumLinkedList.next;
}

const testCases = [
    // Normal
    // Data that is typical (expected) and should be accepted by the system.
    { 
        assert: JSON.stringify(sumLists({ data: 5, next: { data: 2, next: { data: 4, next: null } } }, { data: 3, next: { data: 8, next: { data: 1, next: null } } })),
        expected: JSON.stringify({ data: 8, next: { data: 0, next: { data: 6, next: null } } })
    },
    { 
        assert: JSON.stringify(sumLists({ data: 8, next: { data: 5, next: { data: 7, next: null } } }, { data: 2, next: { data: 9, next: { data: 0, next: null } } })),
        expected: JSON.stringify({ data: 0, next: { data: 5, next: { data: 8, next: null } } })
    },
    { 
        assert: JSON.stringify(sumLists({ data: 9, next: { data: 8, next: { data: 7, next: null } } }, { data: 4, next: { data: 9, next: { data: 0, next: { data: 8, next: null } } } })),
        expected: JSON.stringify({ data: 3, next: { data: 8, next: { data: 8, next: { data: 8, next: null } } } })
    },
    { 
        assert: JSON.stringify(sumLists(
            { data: 5, next: { data: 1, next: { data: 8, next: { data: 9, next: { data: 8, next: null }} } } },
            { data: 4, next: { data: 9, next: { data: 0, next: { data: 8, next: null } } } }
        )),
        expected: JSON.stringify({ data: 9, next: { data: 0, next: { data: 9, next: { data: 7, next: { data: 9, next: null } } } } })
    },

    // Boundary data (extreme data, edge case)
    // Data at the upper or lower limits of expectations that should be accepted by the system.
    { 
        assert: JSON.stringify(sumLists({ data: 1, next: null }, { data: 1, next: null })),
        expected: JSON.stringify({ data: 2, next: null })
    },

    // Abnormal data (erroneous data)
    // Data that falls outside of what is acceptable and should be rejected by the system.
    { assert: sumLists(), expected: false },
    { assert: sumLists(0), expected: false },
    { assert: sumLists(''), expected: false },
    { assert: sumLists([]), expected: false },
    { assert: sumLists({}), expected: false },
    { assert: sumLists(NaN), expected: false },
    { assert: sumLists(null), expected: false },
    { assert: sumLists(false), expected: false },
    { assert: sumLists(new Set()), expected: false },
    { assert: sumLists(new Map()), expected: false },
    { assert: sumLists(undefined), expected: false },
];

// Run tests
testCases.forEach((test, index) => {
    const currentTest = `# Test ${index + 1}`;

    console.group(currentTest);
    console.log('Actual:', test.assert);
    console.log('Expected:', test.expected);
    console.log(test.assert === test.expected ? '🤘 Test PASSED 🤘' : '👎 Test FAILED 👎', '\n');
    console.groupEnd(currentTest);
});
